src/partition.rs

```diff
@@ -0,0 +1,911 @@
+}
+
+/// Represents a memory-mapping of a partition region
+///
+/// Drop this to unmap the memory region
+#[cfg(not(esp_idf_version_major = "4"))]
+pub struct EspMemMappedPartition<'a> {
+    handle: esp_partition_mmap_handle_t,
+    start: usize,
+    _t: core::marker::PhantomData<&'a mut ()>,
+}
+
+#[cfg(not(esp_idf_version_major = "4"))]
+impl EspMemMappedPartition<'_> {
+    /// Returns the start address of the memory-mapped region
+    pub const fn start(&self) -> usize {
+        self.start
+    }
+}
+
+#[cfg(not(esp_idf_version_major = "4"))]
+impl Drop for EspMemMappedPartition<'_> {
+    fn drop(&mut self) {
+        unsafe {
+            esp_partition_munmap(self.handle);
+        }
+    }
+}
+
+/// An iterator over the partitions in the ESP32 flash memory
+pub struct EspPartitionIterator {
+    raw_iter: esp_partition_iterator_t,
+}
+
+impl EspPartitionIterator {
+    /// Create a new partition iterator
+    ///
+    /// # Arguments
+    /// - `partition_type`: The type of partitions to iterate over
+    ///
+    /// # Safety
+    /// Only one partition iterator should be created at a time
+    pub unsafe fn new(partition_type: Option<EspPartitionType>) -> Result<Self, EspError> {
+        let (partition_type, partition_subtype) = partition_type
+            .map(|partition_type| partition_type.raw())
+            .unwrap_or((
+                esp_partition_type_t_ESP_PARTITION_TYPE_ANY,
+                esp_partition_subtype_t_ESP_PARTITION_SUBTYPE_ANY,
+            ));
+
+        let raw_iter = esp_partition_find(partition_type, partition_subtype, core::ptr::null());
+
+        Ok(Self { raw_iter })
+    }
+
+    /// Return the next partition in the iterator
+    pub fn next_partition(&mut self) -> Option<EspPartition> {
+        if self.raw_iter.is_null() {
+            return None;
+        }
+
+        let partition = unsafe { esp_partition_get(self.raw_iter) };
+
+        let value = if partition.is_null() {
+            None
+        } else {
+            Some(unsafe { EspPartition::wrap(partition) })
+        };
+
+        self.raw_iter = unsafe { esp_partition_next(self.raw_iter) };
+
+        value
+    }
+}
+
+impl Drop for EspPartitionIterator {
+    fn drop(&mut self) {
+        unsafe {
+            esp_partition_iterator_release(self.raw_iter);
+        }
+    }
+}
+
+impl Iterator for EspPartitionIterator {
+    type Item = EspPartition;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next_partition()
+    }
+}
+
+/// Represents a partition in the ESP32 flash memory
+#[repr(transparent)]
+pub struct EspPartition(*const esp_partition_t);
+
+impl EspPartition {
+    /// Wraps a raw pointer into a partition
+    ///
+    /// # Safety
+    /// The raw pointer should be a valid one
+    /// It should not be shared in multiple `EspPartition` instances
+    pub unsafe fn wrap(partition: *const esp_partition_t) -> Self {
+        Self(partition)
+    }
+
+    /// Create a new partition by label
```

Comment:
Nit: I think the word "create" is simple to misunderstand in this context. A partition is always created out of band, and not in any way something that this api does. E.g we should drive the point home somewhere ( at the header of this file) that this needs to also happen.

Reply:
Indeed, fixed.